---
title: "Reproduce figures from Leif et al"
author: "Lukas Simon"
date: "27 November 2019"
output: html_document
---

In this analysis we will reproduce figures from (Leif et al)<https://www.cell.com/cell/pdf/S0092-8674(19)30055-8.pdf>.

Load R libraries
```{r warning=FALSE}
library(ComplexHeatmap)
#library(MitoTrace)
source("../R/MitoTrace.R")
```

The BAM files can be downloaded from XXX.
```{r}
bams <- list.files("../data/23_leif_bam/", full.names = T)
bams <- bams[-grep("bai", bams)]
```

Run MitoTrace and calculate allele frequencies (this will take ~ 3min).
```{r warning=FALSE}
mae_res <- MitoTrace(bam_list = bams, fasta = "../data/GRCH38_MT.fa", chr_name = "MT")
al_fre <- calc_allele_frequency(mae_res)
```

Now, we can reproduce the figures from Figure X.
(https://github.com/lkmklsmn/MitoTrace/blob/master/images/reproduce_result.png)
```{}
boxplot(split(af[which(coord_alleles$start == 779 & coord_alleles$altAllele == "C"),], donors == "Donor1_C101"))
Heatmap(target_mutation, cluster_columns = FALSE,col = col_fun,cluster_rows = FALSE,show_column_names = FALSE,name = "Allele\nFrequency", column_title="Donor1 colonies",row_title = "Mitochondiral mutations")
```

