---
title: "MitoTrace works on single cell 10X genomics data"
author: "Mingqiang Wang & Lukas Simon"
date: "December 3, 2019"
output: html_document
---

In this analysis, we demonstrate that MitoTrace could identify the germline mutations on single-cell 10X Genomics data which publicly available from McGinnis et al. (McGinnis et al. 2019). (https://ucsf.app.box.com/s/vg1bycvsjgyg63gkqsputprq5rxzjl6k)

Load R libraries
```{r}
library(data.table)
library(MitoTrace)
library(ggplot2)
```

Read the annotation file
```{r}
demuxlet <- fread("/Users/mwang14/Google Drive/00_Texas_Posdoc_Career/01_Zhao/12_Develop_Mitogenotyping_Tools/21_read_10xGenomics/jurkat/jurkat_293t_demuxlet.best")
bams <- list.files("/Users/mwang14/Google Drive/00_Texas_Posdoc_Career/01_Zhao/12_Develop_Mitogenotyping_Tools/21_read_10xGenomics/jurkat", full.names = T, pattern = ".bam$")
```

Read human GRCH38 Mitochondrial reference genome
```{r}
fasta_loc <- "/Users/mwang14/Google Drive/00_Texas_Posdoc_Career/01_Zhao/12_Develop_Mitogenotyping_Tools/02_GBM_data/GRCH38_MT.fa"
```

MitoTrace calculate the base counts and coverage counts
```{r}
mae_res <- MitoTrace(bam_list = bams, fasta = fasta_loc, chr_name = "MT", min_read = 100)
#mae_res <- MitoTrace(bam_list = bams, fasta = fasta_loc, chr_name = "MT")
```

MitoTrace calculate the allele frequencies
```{r}
af <- calc_allele_frequency(mae_res)
```

Generate plots
```{r}
ok <- intersect(colnames(af), demuxlet$BARCODE)
af <- af[,ok]
cell_line <- sapply(demuxlet$BEST,function(x){strsplit(x,"-")[[1]][[2]]})
cell_line <- cell_line[match(ok, demuxlet$BARCODE)]

good_variants <- names(which(rowMeans(af) > 0.3))
good_variants <- names(tail(sort(apply(af, 1, var)), 20))
af_hv <- data.matrix(af[good_variants, ])
ydata <- Rtsne::Rtsne(t(af_hv))

aframe <- data.frame(cell_line, ydata$Y)
ggplot(aes(X1, X2, color = cell_line), data = aframe) + geom_point()

anno <- data.frame(cell_line)
rownames(anno) <- colnames(af_hv)
pheatmap::pheatmap(af_hv, show_colnames = F, annotation_col = anno)
```
